CMakeLists.txt

```diff
@@ -261,3 +263,95 @@ if(CELERITAS_USE_Doxygen)
 endif()
 
 #----------------------------------------------------------------------------#
+# CONFIG FILE INSTALLATION
+#----------------------------------------------------------------------------#
+
+# Where to configure cmake files
+set(CELER_CMAKE_CONFIG_DIRECTORY "${PROJECT_BINARY_DIR}/share")
+# Where to install configured cmake files
+set(CELER_INSTALL_CMAKECONFIGDIR
+  "${CMAKE_INSTALL_DATAROOTDIR}/cmake/Celeritas")
+
+# Build list of CMake files to install
+set(_cmake_files
+  "${PROJECT_SOURCE_DIR}/cmake/CeleritasGen.cmake"
+  "${PROJECT_SOURCE_DIR}/cmake/CeleritasGen/gen-action.py"
+  "${PROJECT_SOURCE_DIR}/cmake/CeleritasLibrary.cmake"
+)
+foreach(_dep Geant4 HepMC3 ROOT VecGeom)
+  if(CELERITAS_USE_${_dep})
+    list(APPEND _cmake_files "${PROJECT_SOURCE_DIR}/cmake/Find${_dep}.cmake")
+  endif()
+endforeach()
+if(CELERITAS_BUILD_TESTS)
+  list(APPEND _cmake_files
+    "${PROJECT_SOURCE_DIR}/cmake/CeleritasAddTest.cmake"
+  )
+endif()
+
+install(FILES ${_cmake_files}
+  DESTINATION "${CELER_INSTALL_CMAKECONFIGDIR}"
+  COMPONENT devel
+)
+install(DIRECTORY "${PROJECT_SOURCE_DIR}/cmake/backport"
+  DESTINATION "${CELER_INSTALL_CMAKECONFIGDIR}"
+  COMPONENT devel
+)
+
+# Export all cache variables that start with CELERITAS_
+set(CELER_EXPORT_VARIABLES)
+get_directory_property(_cachevar_keys CACHE_VARIABLES)
+foreach(_key IN LISTS _cachevar_keys)
+  if(_key MATCHES "^CELERITAS_")
+    list(APPEND CELER_EXPORT_VARIABLES "set(${_key} \"${${_key}}\")")
+  endif()
+endforeach()
+
+# Add other cache variables, prefixed with CELER_
+foreach(_key BUILD_SHARED_LIBS)
+  list(APPEND CELER_EXPORT_VARIABLES "set(CELERITAS_${_key} \"${${_key}}\")")
+endforeach()
+
+# Add hints for direct dependencies and indirect geant dependencies
+foreach(_key
+  MPIEXEC_EXECUTABLE nlohmann_json_DIR VecCore_DIR VecGeom_DIR Python_DIR Geant4_DIR
+  CLHEP_DIR ZLIB_DIR EXPAT_DIR XercesC_DIR PTL_DIR
+  EXPAT_INCLUDE_DIR EXPAT_LIBRARY
+  XercesC_LIBRARY XercesC_INCLUDE_DIR
+)
+  set(_val "${${_key}}")
+  if(_val)
+    list(APPEND CELER_EXPORT_VARIABLES "set_and_check(${_key} \"${_val}\")")
+  endif()
+endforeach()
```

Comment:
For pure future-proofing (i.e. not critical now), it might be better to put these in a `CeleritasDeps.cmake` file or similar alongside `CeleritasConfig.cmake` and `include` it optionally. That way installations could be made relocatable by _not_ installing that file and relying on `CMAKE_PREFIX_PATH`.

Reply:
Good point. I did this primarily because of a loathsome situation I encounter when building Celeritas against Geant4. When only Geant4 is in the cmake prefix path, `find_package(Geant4)` fail because the expat/xerces/zlib that it was built against (and whose locations are in the RPATHs) aren't being found. It shouldn't be the responsibility of the downstream code to figure out what upstream package dependencies are needed and where they are built...

Should they be `set` instead of `set_and_check`? Should this only be written when  `CMAKE_SKIP_INSTALL_RPATH`  is set? How do we guarantee that the geant4/root/etc. that we find externally have the same configuration that we built with? etc.

I like the `include(OPTIONAL)` idea but it seems pretty unconventional to remove a file to make an installation work with relocations?

Reply:
Yeah, that's the idea with the `Geant4PackageCache.cmake` file in Geant4, but as you found, it only supports two cases (but I'd be interested to know the failure case as Spack at least _should_ set up `CMAKE_PREFIX_PATH`?):

- Use the cache file and give up relocatability
- Don't use the cache file and rely on the system (Spack, etc) to populate `CMAKE_PREFIX_PATH` etc correctly.

There's _probably_ a way to satisfy both, albeit it might end up rather complex (especially if one of the dependencies is found _before_ Geant4/Celeritas). Probably needs more args/additional calls to `find_dependency` to take hints and then fallback gradually.

Given that these keys end up in a single block in the config file, it's probably best to leave this as-is for now and use `set_and_check` as that will at least give a hard error on missing dependencies and allow more time to investigate the issues you noted.

Reply:
FWIW, here's an example of what I mean with Geant4:
```
$ cat CMakeLists.txt
project(find_my_geant LANGUAGES CXX)
cmake_minimum_required(VERSION 3.23)
find_package(Geant4)
$ cmake -DGeant4_ROOT=/rnsdhpc/code/spack/opt/spack/apple-clang-13.0.0/geant4/u3srvfj/ .
-- The CXX compiler identification is AppleClang 13.1.6.13160021
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find CLHEP (missing: CLHEP_DIR)
CMake Warning at CMakeLists.txt:3 (find_package):
  Found package configuration file:

    /rnsdhpc/code/spack/opt/spack/apple-clang-13.0.0/geant4/u3srvfj/lib/Geant4-11.0.0/Geant4Config.cmake

  but it set Geant4_FOUND to FALSE so package "Geant4" is considered to be
  NOT FOUND.  Reason given by package:

  Geant4 could not be found because dependency CLHEP could not be found.



-- Configuring done
-- Generating done
-- Build files have been written to: /Users/s3j/test-geant4-find
```

Reply:
Right, that's indeed the case that the Spack build of Geant4 _doesn't_ install the the "package cache" file: https://github.com/spack/spack/blob/055c9d125d29606947f2fd88a06e279667a74648/var/spack/repos/builtin/packages/geant4/package.py#L199-L202

So an isolated use of this install without the dependent build environment setup (via a module load or Spack environment) wouldn't be expected to find everything - but we can perhaps discuss how to handle this in the Spack Slack?